src/indexer.rs

```diff
@@ -99,45 +107,64 @@ impl Indexer {
         self.inner.clone()
     }
 
+    pub async fn connect_to(&mut self, target: &str) -> Result<()> {
+        info!("Connecting to node for data availability stream");
+        let listener = TcpListener::bind("127.0.0.1:0")
+            .await
+            .expect("Failed to bind to port");
+
+        let addr = listener
+            .local_addr()
+            .expect("Failed to get local port")
+            .to_string();
+
+        let waiter = listener.accept();
+
+        Client::new()
+            .get(format!("{}v1/da/stream_request", target))
+            .query(&[("start_height", "1"), ("peer_ip", &addr)])
+            .timeout(std::time::Duration::from_secs(5))
+            .send()
+            .await?;
+
+        let (stream, _) = waiter.await?;
+        self.da_stream = Some(Framed::new(stream, LengthDelimitedCodec::new()));
+        info!("Connected to data stream");
+
+        Ok(())
+    }
+
     pub async fn start(&mut self) -> Result<()> {
-        handle_messages! {
-            on_bus self.bus,
-            listen<ConsensusEvent> cmd => {
-                self.handle_consensus_event(cmd).await;
+        if self.da_stream.is_none() {
+            handle_messages! {
+                on_bus self.bus,
+                listen<ConsensusEvent> cmd => {
+                    self.handle_consensus_event(cmd).await;
+                }
+            }
+        } else {
+            handle_messages! {
+                    on_bus self.bus,
+                    cmd = self.da_stream.as_mut().unwrap().next() => {
+                    if let Some(Ok(cmd)) = cmd {
+                        let bytes = cmd;
+                        let block: Block = bincode::decode_from_slice(&bytes, bincode::config::standard()).unwrap().0;
+                        self.handle_block(block).await;
+                        SinkExt::<bytes::Bytes>::send(self.da_stream.as_mut().unwrap(), "ok".into()).await.unwrap();
```

Comment:
can we avoid using `unwrap` ? we are in a function that returns a Result<()>, so it should be easy to have a safe code here